### PR TITLE
Vulkan Optimizations and Fixes

### PR DESCRIPTION
**Pull request by 0cc4m**
I have implemented a number of Vulkan optimizations and fixes:
- Implement REPEAT operator shader to fix low performance of Vulkan copy-based implementation
- Use GLSL FMA instruction where possible
- Add GGML_VULKAN_PERF option to get approximate performance data about a running model
- Rework and fix Vulkan Descriptor Set handling, this improves performance in my tests on AMD RADV
- Fix validation error on float32 concat f16 shader

I will keep this on draft while I check a few more things, but feel free to test and benchmark. Don't expect a huge difference.

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

### CODE REVIEW

**Comment by 0cc4m** [comment]
I missed a validation issue in #8943, but the fix is now in this branch. I think this should be ready for a review and then merge.

File: ggml/src/vulkan-shaders/mul_mat_vec_q4_k.comp
Comment (ggerganov) [inline ggml/src/vulkan-shaders/mul_mat_vec_q4_k.comp:103]: If you consider only the FMA changes, is there a measurable performance gain?

Comment (0cc4m) [reply ggml/src/vulkan-shaders/mul_mat_vec_q4_k.comp:103]: It's very hard to tell. The GLSL compiler should be using FMA instructions anyways, basically this change just makes it certain instead of leaving it to the optimizer. Hopefully this means a few more FMA calls in SPIR-V, which could be checked.

But afterwards the SPIR-V code gets compiled again to a device-specific driver-internal representation, where some more optimization takes place. Since there are many combinations of devices, I can't really be sure whether this helped anywhere, but at least I'm sure it doesn't cause slow downs. I haven't seen a significant performance difference on my devices.

**Comment by 0cc4m** [comment]
@ggerganov @slaren Can one of you review the non-Vulkan parts of this PR and approve if that's fine?

**Review by slaren — APPROVED** [review]

**Review by ggerganov — APPROVED** [review]
Make sure to fix the CI before merging

**Pull request merged**
